core/src/oop_module.c

```diff
@@ -0,0 +1,720 @@
+/* SRS_OOP_MODULE_027_038: [Message Channel - If a module message was received, then `OopModule_DoWork` will parse that message by calling `MESSAGE_HANDLE Message_CreateFromByteArray(const unsigned char * source, int32_t size)` with the buffer received from `nn_recv` as `source` and return value from `nn_recv` as `size`] */
+/* SRS_OOP_MODULE_027_039: [Message Channel - If unable to parse the module message, then `OopModule_DoWork` shall free any previously allocated memory and abandon the message channel request] */
+/* SRS_OOP_MODULE_027_040: [Message Channel - `OopModule_DoWork` shall pass the structured message to the module by calling `void Module_Receive(MODULE_HANDLE moduleHandle)` using the parsed message as `moduleHandle`] */
+/* SRS_OOP_MODULE_027_041: [Message Channel - `OopModule_DoWork` shall free the resources held by the parsed module message by calling `void Message_Destroy(MESSAGE_HANDLE * message)` using the parsed module message as `message`] */
+/* SRS_OOP_MODULE_027_042: [Message Channel - `OopModule_DoWork` shall free the resources held by the gateway message by calling `int nn_freemsg(void * msg)` with the resulting buffer from the previous call to `nn_recv`] */
+void
+OopModule_DoWork (
+    OOP_MODULE_HANDLE oop_module
+) {
+    if (NULL == oop_module) {
+        LogError("%s: NULL parameter - oop_module!", __FUNCTION__);
+    } else {
+        int bytes_received;
+        void * control_message = NULL;
+
+        // Check for message on control channel
+        bytes_received = nn_recv(oop_module->control_socket, &control_message, NN_MSG, NN_DONTWAIT);
+        if (EAGAIN == bytes_received) {
+            // no messages available at this time    
+        } else if (0 > bytes_received) {
+            LogError("%s: Unexpected error received from the control channel!", __FUNCTION__);
+        } else {
+            CONTROL_MESSAGE * structured_control_message;
+
+            // Parse message body
+            if (NULL == (structured_control_message = ControlMessage_CreateFromByteArray((const unsigned char *)control_message, bytes_received))) {
+                LogError("%s: Unable to parse control message!", __FUNCTION__);
+            } else {
+                // Route control channel messages to appropriate functions
+                switch (structured_control_message->type) {
+                  case CONTROL_MESSAGE_TYPE_MODULE_CREATE:
+                    if (0 != process_module_create_message(oop_module, (CONTROL_MESSAGE_MODULE_CREATE *)structured_control_message)) {
+                        LogError("%s: Unable to process create message!", __FUNCTION__);
+                    }
+                    break;
+                  case CONTROL_MESSAGE_TYPE_MODULE_START: if (((MODULE_API_1 *)oop_module->module.module_apis)->Module_Start) { ((MODULE_API_1 *)oop_module->module.module_apis)->Module_Start(oop_module->module.module_handle); } break;
```

Comment:
I'm sure they told Picaso that both eyes shouldn't be on the same sides of the head...

fixed.